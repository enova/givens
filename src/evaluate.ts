/* eslint-disable no-param-reassign */

import {
  givenCache,
  givenProps,
  givenTrace,
} from './types';
import GivenError from './givenError';

const evaluate = <T, K extends keyof T>(
  key: K,
  props: givenProps<T>,
  cache: givenCache<T>,
  trace: givenTrace<T>,
  ssi: Function,
): T[K] => {
  let value = cache[key];

  if (value === undefined) {
    // get the most recent declaration for key
    const topFunc = props[key]![props[key]!.length - 1];

    // check for recursive definition, and error if present
    if (trace.includes(key)) {
      const t = trace;
      trace.length = 0; // reset trace

      throw new GivenError(`recursive variable ${key} detected\ntrace: ${t}`, ssi);
    }

    // try to evaluate and cache
    try {
      trace.push(key);
      value = topFunc();
      cache[key] = value;
      trace.pop();
    } catch (err) {
      trace.length = 0;
      throw err;
    }
  }
  return value!;
};

export default evaluate;
