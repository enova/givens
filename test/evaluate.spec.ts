import evaluate from '../src/evaluate';
import {
  givenProps,
  givenCache,
  givenTrace,
} from '../src/types';
import GivenError from '../src/givenError';

jest.unmock('../src/evaluate');
const mockedGivenError = GivenError as unknown as jest.Mock<typeof GivenError>;
mockedGivenError.mockImplementation(
  (message: string) => (new Error(message) as unknown as typeof GivenError),
);

interface TestType {
  key1: string;
  key2: string;
  key3: string;
}

let key: any;
let props: givenProps<TestType>;
let cache: givenCache<TestType>;
let trace: givenTrace<TestType>;
let ssi: Function;

beforeEach(() => {
  key = 'key1';
  props = {
    key1: [() => 'key1 value1', () => 'key1 value2'],
    key2: [() => 'key2 value1', () => 'key2 value2'],
    key3: [() => 'key3 value1'],
  };
  cache = {};
  trace = [];
  ssi = jest.fn();
});

describe('evaluate', () => {
  it('uses the top props', () => {
    expect(evaluate<TestType, 'key1'>(key, props, cache, trace, ssi)).toBe('key1 value2');
  });

  describe('trace detection', () => {
    it('throws the correct error', () => {
      trace = ['key1'];
      const f = () => {
        evaluate<TestType, 'key1'>(key, props, cache, trace, ssi);
      };
      expect(f).toThrowErrorMatchingSnapshot();
    });
    it('shows the correct sequence', () => {
      trace = ['key2', 'key1', 'key3'];
      const f = () => {
        evaluate<TestType, 'key1'>(key, props, cache, trace, ssi);
      };
      expect(f).toThrowErrorMatchingSnapshot();
    });
  });

  describe('error catching', () => {
    it('throws error', () => {
      props.key1 = [() => { throw new Error('error'); }];
      expect(() => evaluate<TestType, 'key1'>(key, props, cache, trace, ssi)).toThrowError('error');
    });

    it('resets trace', () => {
      props.key1 = [() => {
        throw new Error('error');
      }];
      trace = ['key3'];
      expect(() => evaluate<TestType, 'key1'>(key, props, cache, trace, ssi)).toThrowError('error');
      expect(trace).toHaveLength(0);
    });
  });

  describe('value caching', () => {
    it('returns cached value without calling functions', () => {
      const topFn = jest.fn(() => 'value1');
      props.key1 = [topFn];
      cache.key1 = 'cachedValue';
      expect(evaluate<TestType, 'key1'>(key, props, cache, trace, ssi)).toBe('cachedValue');
      expect(topFn).not.toHaveBeenCalled();
    });
  });
});
