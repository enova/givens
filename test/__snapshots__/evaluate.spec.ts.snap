// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluate trace detection shows the correct sequence 1`] = `
"recursive variable key1 detected
trace: key1 => key3 => key1"
`;

exports[`evaluate trace detection throws the correct error 1`] = `
"recursive variable key1 detected
trace: key1 => key1"
`;
